### ISSUE TITLE
Improve accessibility of [math] images generated for EPUB and Digital PDF exports

### ISSUE THREAD

**Issue by SteelWagstaff** [pressbooks/pb-mathjax#128]
At present, Pressbooks export routines for formats that don't support javascript use this repo to generate a PNG or SVG file and insert it into the XHTML file that is used to generate the PDF export. The alt text for these images is the LaTeX input, which is not easy to understand for most users of assistive technology. 

We should produce more meaningful alt text for these images so that they would be read for users of assistive technology in a format that it closer to what they typically hear when encountering mathematics on the web. In addition to generating better alt text, we may want to programmatically add `aria-label` and `data-*` elements to the images. 

## Notes for developers researching this task
This should be possible using the approach documented in the MathJax speech demo repo: https://github.com/mathjax/MathJax-demos-node/tree/master/speech. This demo makes use of this speech rule engine library: https://speechruleengine.org/ + https://github.com/Speech-Rule-Engine/speech-rule-engine

https://docs.mathjax.org/en/latest/basic/accessibility.html#screen-reader-support

In a comment on #39, @beckej13820 suggested that we might attempt to either:
1. > export an image with meaningful MathML [as the alt text, instead of the current LaTeX expression]. This means that we are relying on the end user to have a screenreader that can interpret MathML.

2. > generate meaningful Alt Text. The math to text rules are now directly embedded in MathJax 3.2, and [this github repo](https://github.com/mathjax/MathJax-demos-node/tree/master/speech) has ... worked examples of how developers can call the features in MathJax in specific export routines.

PrinceXML now includes support for using MathJax when rendering a PDF document: https://www.princexml.com/forum/topic/4594/support-for-mathjax-3 + https://github.com/yeslogic/prince-scripts/tree/master/compatibility/MathJax

**Comment by pkra** [comment]
For epub, you might want to consider using mathjax/SRE to populate aria-braillelabel as well.  Once you upgrade to v4, you can even try to basic exploration of SSG content in non-JS environments like epub.

**Comment by beckej13820** [comment]
@pkra If we are making a PDF with embedded math, what's the current best practice?

- SVG image
- Embed the assistive MathML as data-mathml?
- Use the SRE to populate aria-label with a speech string?
- Both?

**Comment by SteelWagstaff** [comment]
We received several suggestions for improving this from Brian Richwine at Indiana University. See https://docs.google.com/document/d/1izaW6PUQGjaBu8XR25LNVWxnrobcPPzSANOg_mpkpX4/edit?usp=sharing

Specific recommendations:
## Provide Output Options for EPUB and PDF exports
1. EPUBs: 
- [ ] Provide a choice between SVG, PNG, or MathML
- [ ] Allow configuring a background color as transparent images with black content often disappear in high contrast settings
2. EPUBs and PDFs:
- [ ] Provide settings to configure Math Speech Grammar for alt-text on images
- [ ] Use MathCAT to generate the speech text; MathJax does not make speech text that is suitable for a non-visual user (it’s too often ambiguous) [this is disputed]
- [ ] Alt-text is plain text, use settings appropriate for plain text speech rendering and non-visual users
- [ ] Set sensible defaults:
    * Generate Speech for: Blindness
    * Speech Style: ClearSpeak 
    * Speech Verbosity: Verbose
    * Speech for chemical equations: Off
## Address rendering issues:
- [ ] Be sensitive to inline vs. displayed equations and render them appropriately
- [ ] Address the fuzziness of bitmap image renderings in EPUBs
- [ ] Address the rendering position of the equation number for numbered equations
- [ ] Fix broken equation reference links in PDFs for digital distribution

**Comment by pkra** [comment]
>  Use MathCAT to generate the speech text; MathJax does not make speech text that is suitable for a non-visual user (it’s too often ambiguous)

@SteelWagstaff that seems like a bold claim. Is there any data to back it up?

**Comment by beckej13820** [comment]
@SteelWagstaff 

I think that there are two main speech engines that MathJax uses: MathSpeak and Clearspeak.

I looked into the history of these speech engines once for a presentation:

- Math Speak was created in the 1950s by the mathematician Nemeth (same creator of Nemeth Braille) as a way to formalize how he would describe equations for the secretaries and typesetters he worked with. This is partly why the "feel" of Math Speak is mostly descriptive, it's describing the typeset equation. x^2 would be described as "x superscript 2." Its a description of the typesetting
- Clear Speak rules created later, and feel more like how a teacher might say it. To me it feels more natural. X^2 would be said as "x squared."

I looked at MathCat for a quick second. And it looks like they default to the Clearspeak rules. So setting sensible defaults in MathJax as you describe in your list of things to do would both be responding to the feedback you got from the other instructor, and keep your scope smaller, because the SRE can use Clearspeak rules directly in MathJax.

I believe you can achieve all of what Brian Richwine is asking for in his documentation within MathJax and it's embedded Speech Rules Engine, without introducing this other tool he mentioned, because you should be able to control both the speech engine and the verbosity level.

**Comment by SteelWagstaff** [comment]
Thanks @pkra and @beckej13820 -- the short answer is I don't know! I'm learning as I go here. I printed the recommendations verbatim as received. Ed -- what you're describing about Clearspeak rules makes sense to me and seems simpler than introducing a yet another math-related dependency I don't fully understand 😂

**Comment by brichwin** [comment]
@beckej13820 @SteelWagstaff 

## Recommended Default Speech Style + Settings
Out of all the speech text style options+settings available via MathCAT (MathCAT ClearSpeak, MathCAT SimpleSpeak) + MathJax 4 (SRE ClearSpeak, SRE MathSpeak), only the **MathCAT ClearSpeak Verbose for Blindness option** creates unambiguous speech text that doesn't depend on / assume any prior training of the listener and isn't outright rejected by most students. 

## Rationale

There are many variables involved in speaking mathematical expressions and the appropriateness of the chosen method depends on the anticipated disability(s) of the audience members, their experience level and preferences, if they want a literal rendering or a semantic interpretation, their learned speech style, etc.

Math speech is linear (1D) and common problem that arises out of that is how to clearly represent math that is 2D (structural). For example. How to disambiguate between these two similar but distinct relatively simple expressions:

<img width="644" height="206" alt="Two math expressions: Expression one: The square root of 4 end root plus 5. Expression two: The square root of 4 plus 5 end root" src="https://github.com/user-attachments/assets/bba98198-b6b6-4dc7-9710-d71265a1d5e8" />

Roughly considered, there are two main audiences for spoken math: 

1. those that can see and are actively looking at the math expression,
2. those that can see and are not looking at the expression or that are visually impaired and depend on speech or braille

A sighted user actively looking at the math expression (the first) has significantly different needs than the second. It's assumed that they can see the expression's dimensionality (structure) and don't need every explicit detail announced to them. This allows for speech that is more like what an instructor would use in a classroom. 

This is apparently what the default settings of ClearSpeak in MathJax/SRE was designed for. MathJax 4.0's default ClearSpeak setting renders both of the above expressions the same ("the square root of 4 plus 5"):

<img width="639" height="334" alt="table showing the latex source for the above expressions and how mathjax renders them both as the square root of 4 plus 5" src="https://github.com/user-attachments/assets/a38ce7f4-7dd0-4966-aadc-fe908e62f069" />

## Not appropriate for visually impaired readers

Consider an EPUB that is being created with images of mathematics (no MathML). A non-sighted user's only resource for accessing the content of the mathematics images will be the images' alt-text values. Thus, that alt-text should be unambiguous.  Alt text is plain text, thus the only mechanism available to do this are to include disambiguating words or pauses (created using punctuation like commas, periods, etc.). Thus, MathJax SRE's ClearSpeak does not meet this need. There are several common patterns where it produces ambiguous speech text.

## MathSpeak is harsh and overly wordy
The other speech style offered by the MathJax SRE is MathSpeak Grammar. As mentioned above, MathSpeak was created by Dr. Nemeth and is pretty much just words substituted for Nemeth Braille code symbols. It was designed to allow creation of unambiguous speech text. At first this sounds promising! However, Dr. Nemeth didn't do user testing - Speech text created using MathSpeak is often overly verbose, harsh sounding, produces cognitive overload, and is generally rejected by students we've worked with. MathSpeak's output becomes extremely verbose once encountering common LaTeX such as aligned equations, matrices, conditional expressions, etc. where it resorts to describing them in a table like structure. I'm showing both the MathSpeak Verbose and the MathSpeak Super Brief settings to demonstrate that the wordiness is present in all  MathSpeak verbosity levels:

<img width="641" height="993" alt="Image" src="https://github.com/user-attachments/assets/ac15e790-b3b8-4666-bd13-e4a28e67084e" />

## MathCAT's ClearSpeak has clear advantages

This is where MathCAT shines as it offers a more nuanced set of speech generation options. For example, for the two square root expressions above, it produces "the square root of 4, end root; plus 5" vs. "the square root of 4 plus 5, end root". Note that if offers both disambiguating speech as well as pauses (via punctuation) to differentiate the "rhythm" of it speech as well.

Notice how gracefully MathCAT's ClearSpeak Verbose both clearly disambiguates expressions and succinctly represents structural content using as natural phrasing as possible:
 
<img width="644" height="605" alt="Image" src="https://github.com/user-attachments/assets/489df648-e744-421c-be4a-e15769db514c" />

**Note:** Only the **Verbose** and **For Blindness** settings of MathCAT ClearSpeak produce guaranteed unambiguous speech text.

## Choices are good though...

User experience is a big part of which math speech style should be used. A student that learned to hear math in MathSpeak Verbose is not going to want to suddenly (without training) here MathCAT's ClearSpeak... and vs. versa. Given that (at IU at least) Pressbooks is used to generate one-off EPUBs for student accommodations having a full suite of speech text generation options is desirable for when MathML is not being generated.

## A note on student preferences

When asked, students will report that they prefer math speech text that generates the fewest words. If not careful in giving them examples of their options, they almost always report that they prefer ClearSpeak terse or SimpleSpeak terse. However, when we give them a quiz to see if they can correctly answer simple math problems by speech text alone, they fail. They start to catch on once we give them enough problems like `\frac{a}{b+c}` vs `\frac{a}{b}+c`, which test their ability to distinguish between common math structures. Only then do they realize and begin to appreciate the necessity of disambiguating wording.

## For advanced listeners: MathCAT SimpleSpeak

The student that already has mastered basic math/algebra and has the patience to then learn and master listening to more advanced forms of speech text could move on to MathCAT's SimpleSpeak. SimpleSpeak is capable of generating unambiguous speech text using even less syllables than MathCAT's ClearSpeak Verbose... but on the assumption that the student knows SimpleSpeak's "trick". That trick is that for "simple terms" disambiguating speech text is not used. 

For example, for the two square root examples above, when:

- the square root is only over the 4 (one term), then the end root is not announced: `the square root of 4; plus 5` 
- only when the square root is over both the 4 and the plus 5, then it will add in the end root: `the square root of 4 plus 5 end root` 

However understanding, learning, and mastering SimpleSpeak's tricks requires already having a mastery of math's dimensionality (structures). It is not appropriate for students just beginning to learn mathematics.

**Comment by brichwin** [comment]
# Case of the disappearing 'a's and why the "eigh"s have it

An eagle eyed reader of my above comment would notice that MathCAT substituted  "eigh" in for the variable 'a' in the speech text. This is because Text-To-Speech engines process alt-text as literary prose and thus use speech rules that affect the pronunciation of pronouns. It's common for the letter a to be spoken diminutively (greatly reduced volume, under-stressed, soft vowel sound) compared to how it would be spoken in a math expression (as a normally stressed, normal volume, hard 'A' sound). For some expressions, the 'a' variable may not be voiced at all!

MathCAT handles this issue (in English text) by substituting in 'eigh' for 'a'. This is another example of a thoughtful treatment that MathCAT performs that MathJax's SRE apparently does not do.

**Comment by brichwin** [comment]
**Note:** You can experiment with both MathCAT's and MathJax's speech text generation here: [Convert AsciiMath, LaTeX or MathML to Speech Text](https://accessibility.iu.edu/tools/math-converter.html).

**Comment by beckej13820** [comment]
@brichwin Thank you for including the explanation as well as the speech text generation tool.

Looking at just one example that you gave, can't we get the same results from MathJax as we do from MathCat?

<img width="769" height="632" alt="Image" src="https://github.com/user-attachments/assets/02224644-48e8-4390-a8ed-63b7cd6f6107" />

Using the converter you provided, this chart compares the output of MathJax Clearspeak explicit (row 4), Clearspeak auto (row 3), MathCAT Clearspeak Verbose (row 2), MathCat Clearspeak Medium (row 1).

It seems to me that MathCat Clearspeak Verbose is the same as MathJax Clearspeak Explicit.

**Comment by brichwin** [comment]
@beckej13820 It might be possible to configure enough manual speech rules to get MathJax ClearSpeak close to MathCAT's. I created the MathJax "ClearSpeak Explicit" adding options to the SRE's domain config. I don't think the options I chose for it completely address the issue. Here are all of the options available (for MathJax 4). You'll notice there are many to chose from!:

| locale | modality | domain     | style |
| ------ | -------- | ---------- | ----- |
| en     | speech   | chromevox  | default |
|        |          | clearspeak | AbsoluteValue_Auto, AbsoluteValue_AbsEnd, AbsoluteValue_Cardinality, AbsoluteValue_Determinant |
|        |          |            | Bar_Auto, Bar_Conjugate |
|        |          |            | Caps_Auto, Caps_SayCaps |
|        |          |            | CombinationPermutation_Auto, CombinationPermutation_ChoosePermute |
|        |          |            | Currency_Auto, Currency_Position, Currency_Prefix |
|        |          |            | Ellipses_Auto, Ellipses_AndSoOn |
|        |          |            | Enclosed_Auto, Enclosed_EndEnclose |
|        |          |            | Exponent_Auto, Exponent_AfterPower, Exponent_Ordinal, Exponent_OrdinalPower |
|        |          |            | Fraction_Auto, Fraction_EndFrac, Fraction_FracOver, Fraction_General, Fraction_GeneralEndFrac, Fraction_Ordinal, Fraction_Over, Fraction_OverEndFrac, Fraction_Per |
|        |          |            | Functions_Auto, Functions_None |
|        |          |            | ImpliedTimes_Auto, ImpliedTimes_MoreImpliedTimes, ImpliedTimes_None |
|        |          |            | Log_Auto, Log_LnAsNaturalLog |
|        |          |            | Matrix_Auto, Matrix_Combinatoric, Matrix_EndMatrix, Matrix_EndVector, Matrix_SilentColNum, Matrix_SpeakColNum, Matrix_Vector |
|        |          |            | MultiLineLabel_Auto, MultiLineLabel_Case, MultiLineLabel_Constraint, MultiLineLabel_Equation, MultiLineLabel_None, MultiLineLabel_Row, MultiLineLabel_Step |
|        |          |            | MultiLineOverview_Auto, MultiLineOverview_None |
|        |          |            | MultiLinePausesBetweenColumns_Auto, MultiLinePausesBetweenColumns_Long, MultiLinePausesBetweenColumns_Short |
|        |          |            | MultsymbolDot_Auto, MultsymbolDot_Dot |
|        |          |            | MultsymbolX_Auto, MultsymbolX_By, MultsymbolX_Cross |
|        |          |            | Paren_Auto, Paren_CoordPoint, Paren_Interval, Paren_Silent, Paren_Speak, Paren_SpeakNestingLevel |
|        |          |            | Prime_Auto, Prime_Angle, Prime_Length |
|        |          |            | Roots_Auto, Roots_PosNegSqRoot, Roots_PosNegSqRootEnd, Roots_RootEnd |
|        |          |            | SetMemberSymbol_Auto, SetMemberSymbol_Belongs, SetMemberSymbol_Element, SetMemberSymbol_Member, SetMemberSymbol_In |
|        |          |            | Sets_Auto, Sets_SilentBracket, Sets_woAll |
|        |          |            | TriangleSymbol_Auto, TriangleSymbol_Delta |
|        |          |            | Trig_Auto, Trig_ArcTrig, Trig_TrigInverse |
|        |          |            | VerticalLine_Auto, VerticalLine_Divides, VerticalLine_Given, VerticalLine_SuchThat |
|        |          | emacspeak  | default |
|        |          | mathspeak  | brief, default, sbrief |
|        |          | default    | alternative, default, literary, physics, plural |
|        | prefix   | default    | default |
|        | summary  | default    | default |
|        |          | mathspeak  | brief, sbrief |

**Note:** You can get these options from the NodeJS version of SRE via `npx sre --opt -P`.

You can experiment with the options by setting the **Speech text style:** setting to **Manual**. The example below shows how to recreate the ClearSpeak Explicit option I created:

<table><tr><td><img  width="772" height="447" alt="Image" src="https://github.com/user-attachments/assets/190ad7a9-eea9-45a9-a729-072b8c0e71ad" /></td></tr></table>

**Comment by beckej13820** [comment]
I just wanted to share the the PDF 2.0 standard now includes MathML, so that MathML can be safely embedded into PDFs to make them accessible. This is part of the PDF-UA2 standard. 

https://www.jsonline.com/press-release/story/159268/accessible-math-comes-to-pdf-pdf-2-0-enables-fully-accessible-stem-content/

The important thing here is it makes the discussion about MathCAT or Clearspeak or MathSpeak a user issue, where they can configure their assistive technology to read according to their preferences.

Lightweight screenreaders that don't interpret MathML will continue to require alt text, but users that use full feature screen readers like NVDA or JAWS should be set. 

So now we should watch release notes from PrinceXML, and look for PDF 2.0 compatibility. (Currently exports up to PDF 1.7 formats)

**Comment by arzola** [comment]
This is great news, thanks for let us know, I'll bring this to our backlog for consideration and explore pdf 2.0

**Comment by brichwin** [comment]
The EPUB format already supports MathML and MathML is the best practice for providing accessible Math in an EPUB on platforms that support it. So, Pressbooks could offer it as an option today:
[https://kb.daisy.org/publishing/docs/html/mathml.html](https://kb.daisy.org/publishing/docs/html/mathml.html)

**Comment by arzola** [comment]
@brichwin thank you definitely adding that also to the backlog, those ugly low quality png images should be replaced soon
